src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,172 @@
+        if (index == 0) {
+            Node<T> newNode = new Node<>(null, head, value);
+            head.prev = newNode;
+            head = newNode;
+        } else {
+            Node<T> currentNode = getNode(index);
+            Node<T> newNode = new Node<>(currentNode.prev, currentNode, value);
+            currentNode.prev.next = newNode;
+            currentNode.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        testIndex(index);
+        Node<T> pointer = getNode(index);
+        if (pointer == null) {
+            return null;
+        } else {
+            return pointer.getValue();
+        }
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        testIndex(index);
+        int count = 0;
+        Node<T> pointer = head;
+        T oldValue = null;
+        while (pointer != null) {
+            if (index == count) {
+                oldValue = pointer.getValue();
+                pointer.setValue(value);
+                break;
+            } else {
+                pointer = pointer.next;
+                count++;
+            }
+        }
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
+        testIndex(index);
+        //   T toReturn = null;
+        Node<T> pointer = getNode(index);
+        if (pointer != null) {
+            T result = pointer.getValue();
+            unlink(pointer);
+            return result;
+        }
         return null;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> pointer = head;
+        while (pointer != null) {
+            if (pointer.getValue() == object || pointer.getValue() != null
+                    && (pointer.getValue().equals(object))) {
+                unlink(pointer);
+                return true;
+            }
+            pointer = pointer.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void unlink(Node<T> node) {
+        if (node.prev == null) {
+            head = node.next;
+        } else {
+            node.prev.next = node.next;
+        }
+        if (node.next == null) {
+            tail = node.prev;
+        } else {
+            node.next.prev = node.prev;
+        }
+        size--;
+    }
+
+    private Node<T> getNode(int index) {
+        int count = 0;
+        Node<T> pointer = head;
+        while (pointer != null) {
+            if (count == index) {
+                return pointer;
+            }
+            pointer = pointer.next;
+            count++;
+        }
+        return null;
+    }
+
+    private void testIndex(int index) {
```

Comment:
test prefix uses in test methods often, use check in regular methods
```suggestion
    private void checkIndex(int index) {
```